machine/machine_core/ssh_connection.py

```diff
@@ -417,15 +417,17 @@ def write(self, dest, content, append=False, owner=None, perm=None):
             append: If True, append to existing file instead of replacing it
             owner: If set, call chown on the file with the given owner string
             perm: Optional file permission as chmod shell string (e.g. "0600")
+
+        The directory of dest is created automatically.
         """
         assert dest
         assert self.ssh_address
 
-        cmd = "cat %s '%s'" % (append and '>>' or '>', dest)
+        cmd = f"mkdir -p '{os.path.dirname(dest)}'; cat {append and '>>' or '>'} '{dest}'"
```

Comment:
Getting a little bit philosophical.  I start to dislike this `;` business.  This is a shells script we're writing, at the end of the day.  Why don't we do something like

```python
cmd = f"""
    mkdir -p {shlex.quote(os.path.dirname(dest))}
    cat {append and '>>' or '>'} {shlex.quote(dest)}
"""
```
(note: if you go the "one block of text" route then `shlex.quote()` in any case, please)

or even go crazy and invent a new mode for m.execute() and do it like

```python
cmd = [
    ['mkdir', '-p', os.path.dirname(dest)],
    ['tee', *['-a'] if append else [], dest]
]
```

Reply:
I have to admit: on first blush I found the nested lists thing to be ridiculous, but the more I think about it, the more I like it.  There's really just no other way to have multiple statements and also get quoting right without thinking about it all the time.

Reply:
As discussed, let's not do this, at least not as part of this PR.